### PR TITLE
Implementation of the http2 observer and decoder.

### PR DESCRIPTION
**Pull request by ganeshmurthy**

### CODE REVIEW

File: src/decoders/http2/http2_decoder.c [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+    return frame_type;
+}
+
+
+static int allocate_move_to_scratch_buffer(qd_decoder_buffer_t  *scratch_buffer, const uint8_t *data, size_t capacity)
+{
+    if (capacity > HTTP2_SCRATCH_BUFFER_MAX_SIZE) {
+        return -1;
+    }
+    scratch_buffer->bytes = malloc(capacity);
```
Comment (kgiusti) [inline src/decoders/http2/http2_decoder.c:149]: Use "qd_malloc" (or better qd_calloc for zeroed memory) - see ctools.h

File: src/decoders/http2/http2_decoder.c [hunk 2]
```diff
@@ -0,0 +1,621 @@
...
+
+
+static int reallocate_move_to_scratch_buffer(qd_decoder_buffer_t  *scratch_buffer, const uint8_t *data, size_t data_length)
+{
+    assert(scratch_buffer->size != 0);
+    size_t old_size = scratch_buffer->size;
+    if (old_size + data_length > HTTP2_SCRATCH_BUFFER_MAX_SIZE) {
+        return -1;
+    }
+    scratch_buffer->bytes = realloc(scratch_buffer->bytes, old_size + data_length);
```
Comment (kgiusti) [inline src/decoders/http2/http2_decoder.c:166]: qd_realloc instead of realloc

File: src/decoders/http2/http2_decoder.c [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+
+/**
+ * A protocol parsing error has occurred.  There is no recovery
+ */
+static void parser_error(qd_http2_decoder_t *decoder, const char *reason)
+{
+    qd_http2_decoder_connection_t *conn_state = decoder->conn_state;
+    qd_log(LOG_HTTP2_DECODER, QD_LOG_ERROR, "[C%"PRIu64"] parser_error - reason=%s", decoder->conn_state->conn_id, reason);
+    if (!conn_state->parse_error) {
+        conn_state->parse_error = strdup(reason);
```
Comment (kgiusti) [inline src/decoders/http2/http2_decoder.c:104]: qd_strdup()

File: src/decoders/http2/http2_decoder.c [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+        }
+    }
+    return 0;
+}
+
+/**
+ * Checks if the nth bit is set in the passed in unsigned byte, data
+ */
+static bool is_nth_bit_set(const uint8_t data, int bit)
+{
```
Comment (kgiusti) [inline src/decoders/http2/http2_decoder.c:293]: assert(bit < 8)

File: src/decoders/http2/http2_decoder.c [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+        scratch_buffer->offset += 1;
+        decoder->frame_length_processed += 1;
+    }
+
+    // We are now finally at a place which matters to us - The Header block fragment. We will look thru and decompress it so we can get the request/response headers.
+    int rv = decompress_headers(decoder, stream_id, scratch_buffer->bytes + scratch_buffer->offset, scratch_buffer->size - scratch_buffer->offset);
+    if(rv < 0) {
+        qd_log(LOG_HTTP2_DECODER, QD_LOG_DEBUG, "[C%"PRIu64"] get_request_headers - failure, moving decoder state to HTTP2_DECODE_ERROR", decoder->conn_state->conn_id);
+        static char error[105];
+        sprintf(error, "get_request_headers - failure, moving decoder state to HTTP2_DECODE_ERROR nghttp2 error code=%i", rv);
```
Comment (kgiusti) [inline src/decoders/http2/http2_decoder.c:339]: snprintf(error, sizeof(error), ...

File: src/decoders/http2/http2_decoder.c (hunk 4 above)
Comment (ganeshmurthy) [reply src/decoders/http2/http2_decoder.c:294]: Done !

**Review by kgiusti — APPROVED** [review]

**Pull request merged**